commons/src/main/java/org/infinispan/commons/util/ConcurrentIntSet.java

```diff
@@ -0,0 +1,527 @@
+package org.infinispan.commons.util;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.NoSuchElementException;
+import java.util.PrimitiveIterator;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerArray;
+import java.util.function.Consumer;
+import java.util.function.IntConsumer;
+import java.util.function.IntPredicate;
+import java.util.function.Predicate;
+import java.util.stream.IntStream;
+import java.util.stream.StreamSupport;
+
+import org.infinispan.commons.io.UnsignedNumeric;
+
+/**
+ * Concurrent implementation of an {@link IntSet}. This implementation is limited in that it can only inserts ints up
+ * to an initialized maximum. Any attempt to insert/remove a larger amount will result in an
+ * {@link IllegalArgumentException} thrown. Note that operations spanning multiple values (ie.
+ * {@link #containsAll(IntSet)}, {@link #removeAll(IntSet)}) are not performed atomically and are done on a per value
+ * basis.
+ * @author wburns
+ * @since 9.3
+ */
+public class ConcurrentIntSet implements InternalIntSet {
+   private final AtomicIntegerArray array;
+
+   // Note per Java Language Specification 15.19 Shift Operators
+   // If the promoted type of the left-hand operand is int, only the five lowest-order bits of the right-hand operand
+   // are used as the shift distance.
+   private static final int ADDRESS_BITS_PER_INT = 5;
+
+   /* Used to shift left or right for a partial int mask */
+   private static final int INT_MASK = 0xffff_ffff;
+
+   private final AtomicInteger currentSize = new AtomicInteger();
+
+   public ConcurrentIntSet(int length) {
```

Comment:
Sure

Reply:
Also unfortunately it isn't really `rangeEndExclusive` because you can insert values higher than it possibly if it isn't a multiple of 32. Maybe something like `minCapacityExclusive`? haha

Reply:
It's definitely not min, because you can't change it later :)

Reply:
well it is min in the idea that this is the minimum capacity it has to support.